<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="UNDR ROVER : Unmapped primer directed ROVER">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>UNDR ROVER</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bjpop/undr_rover">View on GitHub</a>

          <h1 id="project_title">UNDR ROVER</h1>
          <h2 id="project_tagline">Unmapped primer directed ROVER</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bjpop/undr_rover/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bjpop/undr_rover/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h2>
<a id="a-simple-fast-method-of-calling-variants-directly-from-fastqs-without-mapping" class="anchor" href="#a-simple-fast-method-of-calling-variants-directly-from-fastqs-without-mapping" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>A simple, fast method of calling variants directly from FASTQs without mapping</h2>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p>We recommend using virtualenv for installing UNDR ROVER in a sandboxed environment instead of a global install. The instructions below assume you are working in a directory where you can install the software.</p>

<p>UNDR ROVER depends on SciPy, which can be challenging to install using pip. We recommend using your operating system package manner to install SciPy before installing UNDR ROVER. The <code>--system-site-packages</code> option to virtualenv tells pip to use the site-installed version of SciPy.</p>

<pre><code>virtualenv --system-site-packages undr_rover_dev
source undr_rover_dev/bin/activate
pip install git+https://github.com/bjpop/undr_rover
</code></pre>

<p>You can test that the installation worked correctly by asking undr_rover for its version number:</p>

<pre><code>undr_rover --version
undr_rover 2.0.0
</code></pre>

<h3>
<a id="undr-rover-needs-a-human-genome-reference-in-fasta-format" class="anchor" href="#undr-rover-needs-a-human-genome-reference-in-fasta-format" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>UNDR ROVER needs a human genome reference in FASTA format.</h3>

<p>UNDR ROVER requires a human reference genome in a single FASTA format. Your choice of which version to use will depend on other downstream analyses that you might do. HG19 works well if you don't know which one to choose.</p>

<p>You will need to index the reference genome to produce a <code>.fai</code> file. You can do this with samtools:</p>

<pre><code>samtools faidx reference.fasta
</code></pre>

<h2>
<a id="worked-example" class="anchor" href="#worked-example" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Worked example</h2>

<p>The source code of UNDR ROVER comes with a small example dataset for testing purposes.</p>

<h4>
<a id="obtain-a-copy-of-the-undr-rover-source-code" class="anchor" href="#obtain-a-copy-of-the-undr-rover-source-code" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Obtain a copy of the UNDR ROVER source code:</h4>

<pre><code>git clone https://github.com/bjpop/undr_rover
</code></pre>

<p>The example data set is in the directory <code>undr_rover/example</code> within the source code.</p>

<pre><code>cd undr_rover/example/
</code></pre>

<p>As no mapping is undertaken by Undr Rover, this file is required to determine the canonical insert sequence. Note that you will also need the .fai index file. </p>

<h4>
<a id="create-a-file-specifying-target-regions" class="anchor" href="#create-a-file-specifying-target-regions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Create a file specifying target regions.</h4>

<p>The example directory contains one called <code>example_coords.txt</code>.</p>

<p>The file should be tab-separated, and contain information in 5 columns. The first column will contain chromosome numbers, while the second and third columns specify the starting and ending coordinates of the target regions. The fourth and fifth columns will name the forward and reverse primers of each target region. </p>

<p>An example of such a file is shown below:</p>

<pre><code>chr11   31811401    31811498    PAX6_1_F1   PAX6_1_R1
chr11   31811499    31811597    PAX6_1_F2   PAX6_1_R2
chr11   31812210    31812307    PAX6_2_F1   PAX6_2_R1
chr11   31812308    31812406    PAX6_2_F2   PAX6_2_R2
chr11   31812399    31812496    PAX6_2_F3   PAX6_2_R3
chr11   31814945    31815042    PAX6_3_F1   PAX6_3_R1
chr11   31815043    31815140    PAX6_3_F2   PAX6_3_R2
chr11   31815147    31815244    PAX6_4_F1   PAX6_4_R1
chr11   31815242    31815339    PAX6_4_F2   PAX6_4_R2
chr11   31815340    31815437    PAX6_4_F3   PAX6_4_R3
</code></pre>

<p>Specify the name of this file using the <code>--primer_coords</code> command line parameter.</p>

<h4>
<a id="create-a-file-specifying-the-primer-sequences" class="anchor" href="#create-a-file-specifying-the-primer-sequences" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Create a file specifying the primer sequences.</h4>

<p>The example directory contains one called <code>example_primers.txt</code></p>

<p>Undr Rover requires the primer sequences for each primer in order to correctly match reads to target regions. The file should be tab-separated and simply contain 2 columns, with the first containing the primer name and the second containing the primer sequence. </p>

<p>An example of such a file is shown below:</p>

<pre><code>PAX6_1_F1   TAACAGCCATTTTTCTTTCTTTCCT       
PAX6_1_R1   GTGAACCTGATATGTCTCAATACT        
PAX6_1_F2   TTTTTTTTTTTACTGTAATCTTGGCC      
PAX6_1_R2   TGTCTGTTTCTCAAAGGGAATATC        
PAX6_2_F1   GAGATTCTGTTTGGGTAAACTTCT        
PAX6_2_R1   CTACACCCCCCCACATATGCA       
PAX6_2_F2   TGGCTGACTGTTCATGTGTGTC      
PAX6_2_R2   TTTGCCTCTCTCCTCACAGCC       
PAX6_2_F3   AGTATGAGGAGGTCTGGCTGG       
PAX6_2_R3   GGCATCTTTAATGATCAGACTTGT
</code></pre>

<p>Specify the name of this file using the <code>--primer_sequences</code> command line parameter.</p>

<h4>
<a id="run-undr-rover-from-the-command-line" class="anchor" href="#run-undr-rover-from-the-command-line" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Run UNDR ROVER from the command line</h4>

<p>The example directory contains a file called <code>command.sh</code> which contains an illustrative command to test out the program.</p>

<p>The command is:</p>

<pre><code>undr_rover --primer_coords example_coords.txt \
           --primer_sequences example_primers.txt \
           --reference reference.fasta \
           --out example.vcf \
           --genotype \
           example_R1.fastq example_R2.fastq
</code></pre>

<p>If you have FASTQ.gz files, they will need to be decompressed prior to using Undr Rover. </p>

<h4>
<a id="7-a-log-file-is-produced-by-undr-rover" class="anchor" href="#7-a-log-file-is-produced-by-undr-rover" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>7. A log file is produced by Undr Rover.</h4>

<p>A log file with the specified name will be created by Undr Rover. Below is an example of the start of a log file created by Undr Rover. </p>

<pre><code>Mon, 15 Dec 2014 17:04:20 Program started.
Mon, 15 Dec 2014 17:04:20 Command line: undr_rover.py --primer_bases 15 --primer_coords k-plex_roverfile2.txt --primer_sequences k-plex_primers.txt --log log_file --id_info dbsnp.vcf.gz --reference fasta/hg19bis.fa --out variant_calls.vcf --coverdir coverage_files test_fastqs/SD0390_S70_L001_R1_001.fastq test_fastqs/SD0390_S70_L001_R2_001.fastq
Mon, 15 Dec 2014 17:04:21 Processing sample SD0390_S70_L001
Mon, 15 Dec 2014 17:04:22 Processing block chr: chr10, start: 89624214, end: 89624313
Mon, 15 Dec 2014 17:04:22 Number of read pairs in block: 0
Mon, 15 Dec 2014 17:04:22 Number of acceptable read pairs in block: 0
Mon, 15 Dec 2014 17:04:22 Number of variants found in block: 0
Mon, 15 Dec 2014 17:04:22 Processing block chr: chr10, start: 89653675, end: 89653774
Mon, 15 Dec 2014 17:04:22 Number of read pairs in block: 0
Mon, 15 Dec 2014 17:04:22 Number of acceptable read pairs in block: 0
Mon, 15 Dec 2014 17:04:22 Number of variants found in block: 0
</code></pre>

<h4>
<a id="8-example-of-a-complete-command-line-for-undr-rover" class="anchor" href="#8-example-of-a-complete-command-line-for-undr-rover" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>8. Example of a complete command line for Undr Rover.</h4>

<pre><code>undr_rover.py   
    --primer_coords roverfile.txt 
    --primer_sequences primers.txt 
    --log log_file 
    --reference fasta/hg19bis.fa 
    --out variant_calls.vcf 
    --coverdir coverage_files 
    sample1*.fastq sample2*.fastq
</code></pre>

<p>Note that there are various optional command line arguments which can be used to alter the behaviour of Undr Rover. Further details on these can be found in the README on Undr Rover's <a href="https://github.com/rli99/undr_rover">GitHub page</a>.</p>

<p>This command line assumes that there are two FASTQ files each for sample1* and sample2*, which contain the first and second reads respectively for each read-pair. Roverfile.txt and primers.txt are both TSV format files containing data as described earlier. </p>

<p>The variants called will be written to the file variant_calls.vcf. This file will report called variants in accordance with the VCF v4.2 specification outlined <a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf">here</a>. Coverage files containing the number of read-pairs which mapped to each region will be output in coverage_files/sample1.coverage and coverage_files/sample2.coverage. A log file describing the actions taken by the program will be stored in log_file.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">UNDR ROVER maintained by <a href="https://github.com/bjpop">bjpop</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
