<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="UNDR ROVER : Unmapped primer directed ROVER">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>UNDR ROVER</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bjpop/undr_rover">View on GitHub</a>

          <h1 id="project_title">UNDR ROVER</h1>
          <h2 id="project_tagline">Unmapped primer directed ROVER</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bjpop/undr_rover/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bjpop/undr_rover/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h2>
<a id="undr-rover---a-fast-and-accurate-variant-caller-for-targeted-dna-sequencing" class="anchor" href="#undr-rover---a-fast-and-accurate-variant-caller-for-targeted-dna-sequencing" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>UNDR ROVER - A fast and accurate variant caller for targeted DNA sequencing.</h2>

<h3>
<a id="authors" class="anchor" href="#authors" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Authors:</h3>

<ul>
<li>Bernard J Pope (2,3)</li>
<li>TÃº Nguyen-Dumont (1)</li>
<li>Daniel J Park (1)</li>
<li>Roger Li (2)</li>
<li>Edmund Lau (2)</li>
<li>Peter Georgeson (2)</li>
</ul>

<ol>
<li>Genetic Epidemiology Laboratory, Department of Pathology, The University of Melbourne.</li>
<li>Victorian Life Sciences Computation Initiative (VLSCI).</li>
<li>Department of Computing and Information Systems, The University of Melbourne.</li>
</ol>

<h3>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>License:</h3>

<p><a href="http://opensource.org/licenses/BSD-3-Clause">The BSD 3-Clause License</a></p>

<h3>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Requirements</h3>

<ul>
<li>Python 2.7</li>
<li><a href="https://pypi.python.org/pypi/PyVCF">PyVCF</a></li>
<li><a href="https://pypi.python.org/pypi/pyfaidx">Pyfaidx</a></li>
<li><a href="https://pypi.python.org/pypi/biopython">Biopython</a></li>
<li><a href="http://www.scipy.org/">SciPy</a></li>
<li>A reference human genome sequence in a single FASTA file</li>
<li>A file describing the amplicon target coordinates</li>
<li>A file describing the PCR primers used for the specified coordinates</li>
<li>Paired-end DNA sequences where the paired reads significantly overlap, such as the result of the <a href="http://hiplex.org/">Hi-Plex targeted sequencing system</a>.</li>
</ul>

<h3>
<a id="description" class="anchor" href="#description" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Description</h3>

<p>UNDR ROVER enables the user to call variants from decompressed FASTQ files
directly without having to go through a mapping step.</p>

<p>Reads are organised into blocks, which are initialised from information about
known primer sequences. For each block, we record the primer's DNA sequence,
coordinates and reference insert sequence. Once the reads have been assigned
to blocks (by matching the bases in their primer region to a known primer),
we call variants for each block one at a time.</p>

<p>The approach UNDR ROVER takes when processing reads to call variants is to
initially assume that all reads contain only single nucleotide variants. If
we detect two single nucleotide variants in a row during this variant calling,
UNDR ROVER immediatly ceases this step and instead does a full gapped alignment
of the read against the reference insert sequence to detect possible insertions
and deletions.</p>

<p>Only variants detected in both reads of a read-pair are considered relevant. This allows UNDR ROVER to filter out many sequencing artifacts and produce highly accurate results. However, it does depend on the underlying sequences to have substantially overlapping paired reads, such as the result of the <a href="http://hiplex.org/">Hi-Plex targeted sequencing system</a>.
User-defined thresholds determine the minimum number and proportion of
read-pairs a variant must be observed to be considered a 'PASS'. Coverage is
also reported for each block to allow for regions which require additional
screening to be easily identified.</p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p>We recommend using virtualenv for installing UNDR ROVER in a sandboxed environment instead of a global install. The instructions below assume you are working in a directory where you can install the software.</p>

<p>UNDR ROVER depends on SciPy, which can be challenging to install using pip. We recommend using your operating system package manner to install SciPy before installing UNDR ROVER. The <code>--system-site-packages</code> option to virtualenv tells pip to use the site-installed version of SciPy.</p>

<pre><code>virtualenv --system-site-packages undr_rover_dev
source undr_rover_dev/bin/activate
pip install git+https://github.com/bjpop/undr_rover
</code></pre>

<p>You can test that the installation worked correctly by asking undr_rover for its version number:</p>

<pre><code>undr_rover --version
undr_rover 2.0.0
</code></pre>

<h3>
<a id="the-human-genome-reference" class="anchor" href="#the-human-genome-reference" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>The human genome reference</h3>

<p>UNDR ROVER requires a human reference genome in a single FASTA format. Your choice of which version to use will depend on other downstream analyses that you might do. HG19 works well if you don't know which one to choose.</p>

<p>You will need to index the reference genome to produce a <code>.fai</code> file. You can do this with samtools:</p>

<pre><code>samtools faidx reference.fasta
</code></pre>

<h2>
<a id="worked-example" class="anchor" href="#worked-example" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Worked example</h2>

<p>The source code of UNDR ROVER comes with a small example dataset for testing purposes.</p>

<h4>
<a id="obtain-a-copy-of-the-undr-rover-source-code" class="anchor" href="#obtain-a-copy-of-the-undr-rover-source-code" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Obtain a copy of the UNDR ROVER source code:</h4>

<pre><code>git clone https://github.com/bjpop/undr_rover
</code></pre>

<p>The example data set is in the directory <code>undr_rover/example</code> within the source code.</p>

<pre><code>cd undr_rover/example/
</code></pre>

<p>As no mapping is undertaken by UNDR ROVER, this file is required to determine the canonical insert sequence. Note that you will also need the .fai index file. </p>

<h4>
<a id="create-a-file-specifying-target-regions" class="anchor" href="#create-a-file-specifying-target-regions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Create a file specifying target regions.</h4>

<p>The example directory contains one called <code>example_coords.txt</code>.</p>

<p>The file should be tab-separated, and contain information in 5 columns. The first column will contain chromosome numbers, while the second and third columns specify the starting and ending coordinates of the target regions. The fourth and fifth columns will name the forward and reverse primers of each target region. </p>

<p>An example of such a file is shown below:</p>

<pre><code>chr11   31811401    31811498    PAX6_1_F1   PAX6_1_R1
chr11   31811499    31811597    PAX6_1_F2   PAX6_1_R2
chr11   31812210    31812307    PAX6_2_F1   PAX6_2_R1
chr11   31812308    31812406    PAX6_2_F2   PAX6_2_R2
chr11   31812399    31812496    PAX6_2_F3   PAX6_2_R3
chr11   31814945    31815042    PAX6_3_F1   PAX6_3_R1
chr11   31815043    31815140    PAX6_3_F2   PAX6_3_R2
chr11   31815147    31815244    PAX6_4_F1   PAX6_4_R1
chr11   31815242    31815339    PAX6_4_F2   PAX6_4_R2
chr11   31815340    31815437    PAX6_4_F3   PAX6_4_R3
</code></pre>

<p>Specify the name of this file using the <code>--primer_coords</code> command line parameter.</p>

<h4>
<a id="create-a-file-specifying-the-primer-sequences" class="anchor" href="#create-a-file-specifying-the-primer-sequences" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Create a file specifying the primer sequences.</h4>

<p>The example directory contains one called <code>example_primers.txt</code></p>

<p>UNDR ROVER requires the primer sequences for each primer in order to correctly match reads to target regions. The file should be tab-separated and simply contain 2 columns, with the first containing the primer name and the second containing the primer sequence. </p>

<p>An example of such a file is shown below:</p>

<pre><code>PAX6_1_F1   TAACAGCCATTTTTCTTTCTTTCCT       
PAX6_1_R1   GTGAACCTGATATGTCTCAATACT        
PAX6_1_F2   TTTTTTTTTTTACTGTAATCTTGGCC      
PAX6_1_R2   TGTCTGTTTCTCAAAGGGAATATC        
PAX6_2_F1   GAGATTCTGTTTGGGTAAACTTCT        
PAX6_2_R1   CTACACCCCCCCACATATGCA       
PAX6_2_F2   TGGCTGACTGTTCATGTGTGTC      
PAX6_2_R2   TTTGCCTCTCTCCTCACAGCC       
PAX6_2_F3   AGTATGAGGAGGTCTGGCTGG       
PAX6_2_R3   GGCATCTTTAATGATCAGACTTGT
</code></pre>

<p>Specify the name of this file using the <code>--primer_sequences</code> command line parameter.</p>

<h4>
<a id="run-undr-rover-from-the-command-line" class="anchor" href="#run-undr-rover-from-the-command-line" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Run UNDR ROVER from the command line</h4>

<p>The example directory contains a file called <code>command.sh</code> which contains an illustrative command to test out the program.</p>

<p>The command is:</p>

<pre><code>undr_rover --primer_coords example_coords.txt \
           --primer_sequences example_primers.txt \
           --reference reference.fasta \
           --out example.vcf \
           --genotype \
           example_R1.fastq example_R2.fastq
</code></pre>

<p>The output of the above command will be a VCF file called <code>example.vcf</code>, its contents will look like so:</p>

<pre><code>##fileformat=VCFv4.2
##fileDate=20160324
##source=UNDR ROVER
##INFO=&lt;ID=Sample,Number=1,Type=String,Description="Sample Name"&gt;
##INFO=&lt;ID=NV,Number=1,Type=Float,Description="Number of read pairs with variant"&gt;
##INFO=&lt;ID=NP,Number=1,Type=Float,Description="Number of read pairs at POS"&gt;
##INFO=&lt;ID=PCT,Number=1,Type=Float,Description="Percentage of read pairs at POS with variant"&gt;
##INFO=&lt;ID=GT,Number=1,Type=String,Description="Most likely genotype of the variant"&gt;
##INFO=&lt;ID=GTP,Number=1,Type=Float,Description="Probability of the genotype"&gt;
##FILTER=&lt;ID=at,Description="Variant does not appear in at least 2 read pairs"&gt;
##FILTER=&lt;ID=pt,Description="Variant does not appear in at least 5.0% of read pairs for the given region"&gt;
#CHROM  POS ID  REF ALT QUAL    FILTER  INFO
chr7    152357834   .   T   G   .   at;pt Sample=example_R1.fastq;NV=1;NP=155;PCT=0.65%;GT=TT;GTP=0.659
chr7    152357819   .   T   C   .   PASS    Sample=example_R1.fastq;NV=76;NP=155;PCT=49.03%;GT=TC;GTP=0.957
</code></pre>

<p>The VCF file contains a header region which describes the contents of the file, plus one line for each reported variant. Note that not all reported variants will be true. UNDR ROVER reports all sequence differences that it sees in the input sample which appear on both reads in a pair. It then applies stringency filters to those reported variants and annotates them in the FILTER column with PASS if they pass all stringency filters, or with annotations describing which filters they fail (<code>at</code> and <code>pt</code>). In the example above there are two reported variants, although only one of them passes the default filters (PASS). The first reported variant only appears in 1 read pair (NV=1) out of 155 read pairs which aligned to the same position. Therefore it fails two stringency tests: <code>at</code> Variant does not appear in at least 2 read pairs, and <code>pt</code> Variant does not appear in at least 5.0% of read pairs for the given region. These stringency filters are adjustable on the command line. Furthermore we can see that the reference base was T and the alternative (putative variant) base was G, however the genotype called for this position is homozygous for the reference base (GT=TT). This says that UNDR ROVER believes that the evidence in the reads for this position suggests that the reported variant G is an error. UNDR ROVER is quite confident in its reported genotype for this position (GTP=0.659). The second reported variant passes all the stringency tests and agrees with the highly confident computed genotype (GT=TC, GTP=0.957). Out of 155 read pairs for this position, the variant was observed in about half (NV=76).</p>

<p>UNDR ROVER can accept multiple samples on the input command line. You must specify two files for each sample, corresponding to the first and second read in each pair. The read pair files for each sample must be specified consecutively. UNDR ROVER will call the variants in each sample individually, it does not perform joint genotyping.</p>

<p>If you have FASTQ.gz files, they will need to be decompressed. </p>

<p>The variants called will be written to the specified output file. This file will report called variants in accordance with the <a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF v4.2 specification</a>. Coverage files containing the number of read-pairs which mapped to each region will be output in coverage_files/sample1.coverage and coverage_files/sample2.coverage. A log file describing the actions taken by the program will be stored in log_file.</p>

<h3>
<a id="command-line-arguments" class="anchor" href="#command-line-arguments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Command line arguments</h3>

<p>Use the <code>-h</code> command line argument to get a listing of all UNDR ROVER's options:</p>

<pre><code>undr_rover -h
</code></pre>

<p>The above command produces the following output. A more detailed explanation of the options is provided below.</p>

<p>Many of the command line arguments have default values.</p>

<pre><code>usage: undr_rover [-h] [--version] --primer_coords PRIMER_COORDS
                  --primer_sequences FILE
                  [--primer_prefix_size PRIMER_PREFIX_SIZE]
                  [--kmer_length KMER_LENGTH]
                  [--kmer_threshold KMER_THRESHOLD]
                  [--primer_bases PRIMER_BASES] [--proportionthresh N]
                  [--absthresh N] [--qualthresh N] [--overlap OVERLAP]
                  [--max_variants N] --reference FILE [--id_info FILE] --out
                  FILE [--log FILE] [--coverdir COVERDIR] [--fast]
                  [--snvthresh N] [--genotype] [--ploidy {1,2}]
                  fastqs [fastqs ...]

Find variants from fastqs via a mapping-free approach.

positional arguments:
  fastqs                Fastq files containing reads.

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  --primer_coords PRIMER_COORDS
                        Primer coordinates in TSV format.
  --primer_sequences FILE
                        Primer base sequences as determined by a primer
                        generating program.
  --primer_prefix_size PRIMER_PREFIX_SIZE
                        Size of primer key for blocks in dictionary.
  --kmer_length KMER_LENGTH
                        Length of k-mer to check after the primer sequence.
                        Defaults to 30.
  --kmer_threshold KMER_THRESHOLD
                        Number of single nucleotide variants deemed acceptable
                        in kmer.
  --primer_bases PRIMER_BASES
                        Number of bases from primer region to use in gapped
                        alignment.Helps with variant calling near the edges of
                        a block. Defaults to 5.
  --proportionthresh N  Keep variants which appear in this proportion of the
                        read pairs. For a given target region, and bin
                        otherwise. Defaults to 0.05.
  --absthresh N         Only keep variants which appear in at least this many
                        read pairs. Defaults to 2.
  --qualthresh N        Minimum base quality score (phred).
  --overlap OVERLAP     Minimum proportion of block which must be overlapped
                        by a read. Defaults to 0.9.
  --max_variants N      Ignore reads with greater than this many variants
                        observed. Defaults to 20.
  --reference FILE      Reference sequences in Fasta format.
  --id_info FILE        File containing rs ID information in VCF format.
  --out FILE            Name of output file containing called variants.
  --log FILE            Logs progress in specified file, defaults to stdout.
  --coverdir COVERDIR   Directory to write coverage files, defaults to current
                        working directory.
  --fast                Use gapped alignment less often, leading to faster run
                        time.
  --snvthresh N         Distance between two single nucleotide variants before
                        going to a gapped alignment.
  --genotype            Compute genotypes for SNVs. Defaults to False.
  --ploidy {1,2}        Ploidy for genotyping 1 = haploid, 2 = diploid.
                        Defaults to 2.
</code></pre>

<h3>
<a id="explanation-of-command-line-options" class="anchor" href="#explanation-of-command-line-options" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Explanation of command line options</h3>

<h4>
<a id="-h" class="anchor" href="#-h" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>-h</code>
</h4>

<p>Print a help message and exit.</p>

<h4>
<a id="--version" class="anchor" href="#--version" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--version</code>
</h4>

<p>Print the version number of rover and exit.</p>

<h4>
<a id="--primer_coords-coords" class="anchor" href="#--primer_coords-coords" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--primer_coords COORDS</code>
</h4>

<p><strong>Required</strong></p>

<p>A list of primers with their expected coordinates. TSV format with the following data:</p>

<pre><code>chromosome  start   end     forward_primer  reverse_primer
</code></pre>

<h4>
<a id="--primer_sequences-seq" class="anchor" href="#--primer_sequences-seq" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--primer_sequences SEQ</code>
</h4>

<p><strong>Required</strong></p>

<p>A list of primers with their expected base sequences. TSV format with
the following data:</p>

<pre><code>primer_name sequence
</code></pre>

<h4>
<a id="--primer_prefix_size-n" class="anchor" href="#--primer_prefix_size-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--primer_prefix_size N</code>
</h4>

<p><strong>Optional</strong>. Defaults to 20. </p>

<p>Determines the size of the prefix used for the dictionary which stores 
primers. Will most likely be set to a value equal to the minimum length 
of a primer. </p>

<h4>
<a id="--kmer_length-n" class="anchor" href="#--kmer_length-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--kmer_length N</code>
</h4>

<p><strong>Optional</strong>. Defaults to 30.</p>

<p>Determines the length of k-mer to use in k-mer test. In this 
test, if a certain number of SNVs is found in k-mers from both reads
then the read pair is discarded.</p>

<h4>
<a id="--kmer_threshold-n" class="anchor" href="#--kmer_threshold-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--kmer_threshold N</code>
</h4>

<p><strong>Optional</strong>. Defaults to 2.</p>

<p>The number of SNVs considered acceptable in the k-mer region of a read.
If there are more than N SNVs in both k-mers of a read pair, then the 
read will be discarded. </p>

<h4>
<a id="--primer_bases-n" class="anchor" href="#--primer_bases-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--primer_bases N</code>
</h4>

<p><strong>Optional</strong>. Defaults to 3.</p>

<p>When there are repeated bases or repeated sequences of bases near the 
start of a block, detecting variants becomes slightly harder as some 
variants won't be apparent. By default, UNDR ROVER will use 3 additional 
bases from the end of the primer region during variant calling. This will 
help to detect those variants. </p>

<p>Increasing the value of this parameter will mean more bases from the end 
of the primer region are used, leading to slightly improved accuracy at 
the cost of slower run time.  </p>

<h4>
<a id="--proportionthresh-n" class="anchor" href="#--proportionthresh-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--proportionthresh N</code>
</h4>

<p><strong>Optional</strong>. Defaults to 0.05.</p>

<p>Only keep variants which appear in this proportion of the read-pairs for
a given target region, and bin otherwise. A variant must appear in
both reads of a pair to be counted. The proportion is calculated as
follows:</p>

<ul>
<li>N = number of pairs containing this variant in both reads</li>
<li>T = number of read-pairs overlapping the target region</li>
</ul>

<p>proportion = N/T</p>

<p>Note: variants must pass BOTH the proportionthresh and absthresh 
thresholds to be kept. If they fail either test then they are discarded.</p>

<p>That is to say:</p>

<pre><code>if N/T  &gt;= proportionthresh and N &gt;= absthresh:
    keep the variant
else:
    discard the variant
</code></pre>

<h4>
<a id="--absthresh-n" class="anchor" href="#--absthresh-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--absthresh N</code>
</h4>

<p><strong>Optional</strong>. Defaults to 2 read-pairs.</p>

<p>Only keep variants which appear in at least this many read-pairs
for a given target region.</p>

<p>See comments above about <code>proportionthresh</code>.</p>

<h4>
<a id="--qualthresh-n" class="anchor" href="#--qualthresh-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--qualthresh N</code>
</h4>

<p><strong>Optional</strong>.</p>

<p>Minimum phred quality score for bases appearing in SNVs and
insertions. If this argument is set Rover will only consider SNVs and
insertions where all DNA bases in those variants have a quality score
greater than or equal to the argument. For example, setting this
argument to 35 will cause Rover to discard any SNVs or insertions
containing any bases with a score below 35. If the argument is not
set then Rover will not consider quality scores in its decision
to keep or discard a variant.</p>

<h4>
<a id="--overlap-overlap" class="anchor" href="#--overlap-overlap" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--overlap OVERLAP</code>
</h4>

<p><strong>Optional</strong>. Defaults to 0.9.</p>

<p>Minimum fraction overlap of read to block region.
0.5 means at least half of a block must be overlapped by a read
for the read to be considered for that region. 1.0 would mean the entire
block must be overlapped by the read.</p>

<h4>
<a id="--max_variants-n" class="anchor" href="#--max_variants-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--max_variants N</code>
</h4>

<p><strong>Optional</strong>. Defaults to 25. </p>

<p>The maximum number of variants which can be observed for a single read
before it is considered meaningless and discarded for the purposes of
variant calling. </p>

<h4>
<a id="--reference-file" class="anchor" href="#--reference-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--reference FILE</code>
</h4>

<p><strong>Required</strong>.</p>

<p>Reference sequences in FASTA format. UNDR ROVER extracts insert 
sequences for each block from the reference and compares this with the
read sequences to call variants.</p>

<h4>
<a id="--id_info-file" class="anchor" href="#--id_info-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--id_info FILE</code>
</h4>

<p><strong>Optional</strong>.</p>

<p>UNDR ROVER will find the rs numbers in DBSNP and show them in the 
output VCF file if it can find the relevant number in DBSNP, which is a 
vcf format file containing rs numbers for known variants in .vcf.gz 
format with an accompanying .tbi file (created by tabix).</p>

<h4>
<a id="--fast" class="anchor" href="#--fast" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--fast</code>
</h4>

<p><strong>Optional</strong>. Defaults to False.</p>

<p>If set, gapped alignment will only be used when SNVs are detected within 
snvthresh bases of each other. Leads to a significant decrease in the run 
time of UNDR ROVER. </p>

<h4>
<a id="--snvthresh-n" class="anchor" href="#--snvthresh-n" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--snvthresh N</code>
</h4>

<p><strong>Optional</strong>. Defaults to 1.</p>

<p>If <code>--fast</code> is set, gapped alignment will be performed on reads
which contain SNVs within N bases of each other. The default setting
of 1 allows for the fastest run-time and should provide a sufficient 
level of accuracy in the majority of cases.</p>

<h4>
<a id="--coverdir-coverdir" class="anchor" href="#--coverdir-coverdir" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>--coverdir COVERDIR</code>
</h4>

<p><strong>Optional</strong>. Defaults to current working directory.</p>

<p>Directory to write the coverage files. If the directory does not
exist Rover will not try to create it. You must create the directory
yourself.</p>

<p>The format of the coverage files is a TSV with:</p>

<pre><code>chr     block_start     block_end       num_pairs
</code></pre>

<h4>
<a id="fastqs-fastqs-" class="anchor" href="#fastqs-fastqs-" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><code>fastqs [fastqs ...]</code>
</h4>

<p>One or more pairs of FASTQ files containing reads for which variant
calling will be attempted. FASTQ.gz files will need to decompressed
prior to use with UNDR ROVER.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">UNDR ROVER maintained by <a href="https://github.com/bjpop">bjpop</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
