<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="UNDR ROVER : Unmapped primer directed ROVER">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>UNDR ROVER</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bjpop/undr_rover">View on GitHub</a>

          <h1 id="project_title">UNDR ROVER</h1>
          <h2 id="project_tagline">Unmapped primer directed ROVER</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bjpop/undr_rover/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bjpop/undr_rover/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h2>
<a id="a-simple-fast-method-of-calling-variants-directly-from-fastqs-without-mapping" class="anchor" href="#a-simple-fast-method-of-calling-variants-directly-from-fastqs-without-mapping" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>A simple, fast method of calling variants directly from FASTQs without mapping</h2>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p>We recommend using virtualenv for installing UNDR ROVER in a sandboxed environment instead of a global install. The instructions below assume you are working in a directory where you can install the software.</p>

<p>UNDR ROVER depends on SciPy, which can be challenging to install using pip. We recommend using your operating system package manner to install SciPy before installing UNDR ROVER. The <code>--system-site-packages</code> option to virtualenv tells pip to use the site-installed version of SciPy.</p>

<pre><code>virtualenv --system-site-packages undr_rover_dev
source undr_rover_dev/bin/activate
pip install git+https://github.com/bjpop/undr_rover
</code></pre>

<p>You can test that the installation worked correctly by asking undr_rover for its version number:</p>

<pre><code>undr_rover --version
undr_rover 2.0.0
</code></pre>

<h3>
<a id="undr-rover-needs-a-human-genome-reference-in-fasta-format" class="anchor" href="#undr-rover-needs-a-human-genome-reference-in-fasta-format" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>UNDR ROVER needs a human genome reference in FASTA format.</h3>

<p>UNDR ROVER requires a human reference genome in a single FASTA format. Your choice of which version to use will depend on other downstream analyses that you might do. HG19 works well if you don't know which one to choose.</p>

<p>You will need to index the reference genome to produce a <code>.fai</code> file. You can do this with samtools:</p>

<pre><code>samtools faidx reference.fasta
</code></pre>

<h2>
<a id="worked-example" class="anchor" href="#worked-example" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Worked example</h2>

<p>The source code of UNDR ROVER comes with a small example dataset for testing purposes.</p>

<h4>
<a id="obtain-a-copy-of-the-undr-rover-source-code" class="anchor" href="#obtain-a-copy-of-the-undr-rover-source-code" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Obtain a copy of the UNDR ROVER source code:</h4>

<pre><code>git clone https://github.com/bjpop/undr_rover
</code></pre>

<p>The example data set is in the directory <code>undr_rover/example</code> within the source code.</p>

<pre><code>cd undr_rover/example/
</code></pre>

<p>As no mapping is undertaken by Undr Rover, this file is required to determine the canonical insert sequence. Note that you will also need the .fai index file. </p>

<h4>
<a id="create-a-file-specifying-target-regions" class="anchor" href="#create-a-file-specifying-target-regions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Create a file specifying target regions.</h4>

<p>The example directory contains one called <code>example_coords.txt</code>.</p>

<p>The file should be tab-separated, and contain information in 5 columns. The first column will contain chromosome numbers, while the second and third columns specify the starting and ending coordinates of the target regions. The fourth and fifth columns will name the forward and reverse primers of each target region. </p>

<p>An example of such a file is shown below:</p>

<pre><code>chr11   31811401    31811498    PAX6_1_F1   PAX6_1_R1
chr11   31811499    31811597    PAX6_1_F2   PAX6_1_R2
chr11   31812210    31812307    PAX6_2_F1   PAX6_2_R1
chr11   31812308    31812406    PAX6_2_F2   PAX6_2_R2
chr11   31812399    31812496    PAX6_2_F3   PAX6_2_R3
chr11   31814945    31815042    PAX6_3_F1   PAX6_3_R1
chr11   31815043    31815140    PAX6_3_F2   PAX6_3_R2
chr11   31815147    31815244    PAX6_4_F1   PAX6_4_R1
chr11   31815242    31815339    PAX6_4_F2   PAX6_4_R2
chr11   31815340    31815437    PAX6_4_F3   PAX6_4_R3
</code></pre>

<p>Specify the name of this file using the <code>--primer_coords</code> command line parameter.</p>

<h4>
<a id="create-a-file-specifying-the-primer-sequences" class="anchor" href="#create-a-file-specifying-the-primer-sequences" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Create a file specifying the primer sequences.</h4>

<p>The example directory contains one called <code>example_primers.txt</code></p>

<p>Undr Rover requires the primer sequences for each primer in order to correctly match reads to target regions. The file should be tab-separated and simply contain 2 columns, with the first containing the primer name and the second containing the primer sequence. </p>

<p>An example of such a file is shown below:</p>

<pre><code>PAX6_1_F1   TAACAGCCATTTTTCTTTCTTTCCT       
PAX6_1_R1   GTGAACCTGATATGTCTCAATACT        
PAX6_1_F2   TTTTTTTTTTTACTGTAATCTTGGCC      
PAX6_1_R2   TGTCTGTTTCTCAAAGGGAATATC        
PAX6_2_F1   GAGATTCTGTTTGGGTAAACTTCT        
PAX6_2_R1   CTACACCCCCCCACATATGCA       
PAX6_2_F2   TGGCTGACTGTTCATGTGTGTC      
PAX6_2_R2   TTTGCCTCTCTCCTCACAGCC       
PAX6_2_F3   AGTATGAGGAGGTCTGGCTGG       
PAX6_2_R3   GGCATCTTTAATGATCAGACTTGT
</code></pre>

<p>Specify the name of this file using the <code>--primer_sequences</code> command line parameter.</p>

<h4>
<a id="run-undr-rover-from-the-command-line" class="anchor" href="#run-undr-rover-from-the-command-line" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Run UNDR ROVER from the command line</h4>

<p>The example directory contains a file called <code>command.sh</code> which contains an illustrative command to test out the program.</p>

<p>The command is:</p>

<pre><code>undr_rover --primer_coords example_coords.txt \
           --primer_sequences example_primers.txt \
           --reference reference.fasta \
           --out example.vcf \
           --genotype \
           example_R1.fastq example_R2.fastq
</code></pre>

<p>The output of the above command will be a VCF file called <code>example.vcf</code>, its contents will look like so:</p>

<pre><code>##fileformat=VCFv4.2
##fileDate=20160324
##source=UNDR ROVER
##INFO=&lt;ID=Sample,Number=1,Type=String,Description="Sample Name"&gt;
##INFO=&lt;ID=NV,Number=1,Type=Float,Description="Number of read pairs with variant"&gt;
##INFO=&lt;ID=NP,Number=1,Type=Float,Description="Number of read pairs at POS"&gt;
##INFO=&lt;ID=PCT,Number=1,Type=Float,Description="Percentage of read pairs at POS with variant"&gt;
##INFO=&lt;ID=GT,Number=1,Type=String,Description="Most likely genotype of the variant"&gt;
##INFO=&lt;ID=GTP,Number=1,Type=Float,Description="Probability of the genotype"&gt;
##FILTER=&lt;ID=at,Description="Variant does not appear in at least 2 read pairs"&gt;
##FILTER=&lt;ID=pt,Description="Variant does not appear in at least 5.0% of read pairs for the given region"&gt;
#CHROM  POS ID  REF ALT QUAL    FILTER  INFO
chr7    152357834   .   T   G   .   at;pt   Sample=example_R1.fastq;NV=1;NP=155;PCT=0.65%;GT=TT;GTP=0.659
chr7    152357819   .   T   C   .   PASS    Sample=example_R1.fastq;NV=76;NP=155;PCT=49.03%;GT=TC;GTP=0.957
</code></pre>

<p>If you have FASTQ.gz files, they will need to be decompressed prior to using Undr Rover. </p>

<p>The variants called will be written to the file variant_calls.vcf. This file will report called variants in accordance with the VCF v4.2 specification outlined <a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf">here</a>. Coverage files containing the number of read-pairs which mapped to each region will be output in coverage_files/sample1.coverage and coverage_files/sample2.coverage. A log file describing the actions taken by the program will be stored in log_file.</p>

<h3>
<a id="command-line-arguments" class="anchor" href="#command-line-arguments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Command line arguments</h3>

<pre><code>    -h

        Print a help message and exit.

    --version

        Print the version number of rover and exit.

    --primer_coords COORDS

        Required.

        A list of primers with their expected coordinates. TSV format with the
        following data:

            chromosome  start   end     forward_primer  reverse_primer

    --primer_sequences SEQ

        Required.

        A list of primers with their expected base sequences. TSV format with
        the following data:

            primer_name sequence

    --primer_prefix_size N

        Optional. Defaults to 20. 

        Determines the size of the prefix used for the dictionary which stores 
        primers. Will most likely be set to a value equal to the minimum length 
        of a primer. 

    --kmer_length N

        Optional. Defaults to 30.

        Determines the length of k-mer to use in k-mer test. In this 
        test, if a certain number of SNVs is found in k-mers from both reads
        then the read pair is discarded.

    --kmer_threshold N

        Optional. Defaults to 2.

        The number of SNVs considered acceptable in the k-mer region of a read.
        If there are more than N SNVs in both k-mers of a read pair, then the 
        read will be discarded. 

    --primer_bases N

        Optional. Defaults to 3.

        When there are repeated bases or repeated sequences of bases near the 
        start of a block, detecting variants becomes slightly harder as some 
        variants won't be apparent. By default, Undr Rover will use 3 additional 
        bases from the end of the primer region during variant calling. This will 
        help to detect those variants. 

        Increasing the value of this parameter will mean more bases from the end 
        of the primer region are used, leading to slightly improved accuracy at 
        the cost of slower run time.  

    --proportionthresh N

        Optional. Defaults to 0.05.

        Only keep variants which appear in this proportion of the read-pairs for
        a given target region, and bin otherwise. A variant must appear in
        both reads of a pair to be counted. The proportion is calculated as
        follows:

            N = number of pairs containing this variant in both reads
            T = number of read-pairs overlapping the target region

            proportion = N/T

        Note: variants must pass BOTH the proportionthresh and absthresh 
        thresholds to be kept. If they fail either test then they are binned.

        That is to say:

            if N/T  &gt;= proportionthresh and N &gt;= absthresh:
                keep the variant
            else:
                bin the variant

    --absthresh N

        Optional. Defaults to 2 read-pairs.

        Only keep variants which appear in at least this many read-pairs
        for a given target region.

        See comments above about proportionthresh.

    --qualthresh N

        Optional. Minimum phred quality score for bases appearing in SNVs and
        insertions. If this argument is set Rover will only consider SNVs and
        insertions where all DNA bases in those variants have a quality score
        greater than or equal to the argument. For example, setting this
        argument to 35 will cause Rover to discard any SNVs or insertions
        containing any bases with a score below 35. If the argument is not
        set then Rover will not consider quality scores in its decision
        to keep or discard a variant.

    --overlap OVERLAP

       Optional. Defaults to 0.9.

       Minimum fraction overlap of read to block region.
       0.5 means at least half of a block must be overlapped by a read
       for the read to be considered for that region. 1.0 would mean the entire
       block must be overlapped by the read.

    --max_variants N

        Optional. Defaults to 25. 

        The maximum number of variants which can be observed for a single read
        before it is considered meaningless and discarded for the purposes of
        variant calling. 

    --reference FILE

        Required.

        Reference sequences in FASTA format. Undr Rover extracts insert 
        sequences for each block from the reference and compares this with the
        read sequences to call variants.

    --id_info FILE

        Optional. Rover will find the rs numbers in DBSNP and show them in the 
        output VCF file if it can find the relevant number in DBSNP, which is a 
        vcf format file containing rs numbers for known variants in .vcf.gz 
        format with an accompanying .tbi file (created by tabix).

    --fast

        Optional. Defaults to False.

        If set, gapped alignment will only be used when SNVs are detected within 
        snvthresh bases of each other. Leads to a significant decrease in the run 
        time of Undr Rover. 

    --snvthresh N

        Optional. Defaults to 1.

        If --fast is set, gapped alignment will be performed on reads
        which contain SNVs within N bases of each other. The default setting
        of 1 allows for the fastest run-time and should provide a sufficient 
        level of accuracy in the majority of cases.

    --coverdir COVERDIR

        Optional. Defaults to current working directory.

        Directory to write the coverage files. If the directory does not
        exist Rover will not try to create it. You must create the directory
        yourself.

        The format of the coverage files is a TSV with:

        chr     block_start     block_end       num_pairs

    fastqs [fastqs ...]

        One or more pairs of FASTQ files containing reads for which variant
        calling will be attempted. FASTQ.gz files will need to decompressed
        prior to use with Undr Rover.
</code></pre>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">UNDR ROVER maintained by <a href="https://github.com/bjpop">bjpop</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
